core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -91,7 +90,12 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
   /** List all the files currently stored on disk by the disk manager. */
   def getAllFiles(): Seq[File] = {
     // Get all the files inside the array of array of directories
-    subDirs.flatten.filter(_ != null).flatMap { dir =>
+    subDirs.flatMap { dir =>
```

Comment:
No, it has nothing to do with when the update is visible; `volatile` does not matter. The assignment `a = 0x1234` cannot happen until `0x1234` points to the string `"foobar"`. That is the point. Otherwise the JVM completely violates program order. If what you say is true, then all kinds of insane things could happen in the JVM, like dereferencing an object reference before its constructor has even started.